core/src/main/scala/caliban/execution/Executor.scala

```diff
@@ -72,20 +76,38 @@ object Executor {
             Types.listOf(currentField.fieldType).fold(false)(_.isNullable)
           )
         case ObjectStep(objectName, fields) =>
-          val filteredFields = mergeFields(currentField, objectName)
-          val items          = filteredFields.map {
-            case f @ Field(name @ "__typename", _, _, alias, _, _, _, directives, _, _) =>
-              (alias.getOrElse(name), PureStep(StringValue(objectName)), fieldInfo(f, path, directives))
-            case f @ Field(name, _, _, alias, _, _, args, directives, _, _)             =>
-              (
-                alias.getOrElse(name),
-                fields
-                  .get(name)
-                  .fold(NullStep: ReducedStep[R])(reduceStep(_, f, args, Left(alias.getOrElse(name)) :: path)),
-                fieldInfo(f, path, directives)
+          val filteredFields    = mergeFields(currentField, objectName)
+          val (deferred, eager) = filteredFields.partitionMap {
+            case f @ Field(name @ "__typename", _, _, alias, _, _, _, directives, _, _, _) =>
+              Right((alias.getOrElse(name), PureStep(StringValue(objectName)), fieldInfo(f, path, directives)))
+            case f @ Field(name, _, _, alias, _, _, args, directives, _, _, fragment)      =>
+              val aliasedName = alias.getOrElse(name)
+              val field       = fields
+                .get(name)
+                .fold(NullStep: ReducedStep[R])(reduceStep(_, f, args, Left(alias.getOrElse(name)) :: path))
+
+              val info = fieldInfo(f, path, directives)
+
+              fragment.collectFirst {
+                // The defer spec provides some latitude on how we handle responses. Since it is more performant to return
+                // pure fields rather than spin up the defer machinery we return pure fields immediately to the caller.
+                case IsDeferred(label) if featureSet(Feature.Defer) && !field.isPure =>
```

Comment:
You don't use `Feature.Stream` anywhere, is that still relevant?

Reply:
No I can remove it, originally I had both deferred and streaming in this PR but the streaming changes became much more complicated so I meant to split those out. Technically I could keep the directive support and just not actually provide streaming yet, the spec allows for the server to choose _not_ to stream the values if a single value would be more efficient.